### PR TITLE
Add ability to pass a custom message format

### PR DESCRIPTION
**Pull request by jzoldak**
Including job name and parameters for triggering.

### CODE REVIEW

File: src/main/java/com/base2services/jenkins/SqsQueueHandler.java [hunk 1]
```diff
@@ -10,6 +10,7 @@
 import hudson.util.SequentialExecutionQueue;
 import hudson.util.TimeUnit2;
 
+import java.util.ArrayList;
```
Comment (jzoldak) [inline src/main/java/com/base2services/jenkins/SqsQueueHandler.java]: Reviewer note: 
Changes to this file are unrelated to the new feature. They are improvements to the existing code that makes troubleshooting easier, especially AWS permissions issues. I should probably put them in a separate commit and put in an upstream PR, but haven't gotten around to doing that.

File: src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java [hunk 2]
```diff
@@ -71,8 +101,101 @@ public void processGitHubPayload(String payload, Class<? extends Trigger> trigge
...
+                    }
+                }
+            } catch (JSONException e) {
+                LOGGER.warning("Parameters must be passed as a JSONArray in the SQS message.");
+            }
+        }
+        return params;
+    }
+
+    public JSONObject extractJsonFromPayload(String payload) {
```
Comment (jzoldak) [inline src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java]: Reviewer note:
Changed this function to public so that it can be used in tests.
Also changed its functionality to be more generic - as written it would not extract the payload if there was no "repository" object in it but instead would return null.

File: src/test/java/com/base2services/jenkins/GitHubTriggerProcessorTest.java [hunk 3]
```diff
@@ -72,7 +72,7 @@ public void shouldNotTriggerWithAFork() throws Exception {
     public void shouldNotTriggerWithBadPayload() throws Exception {
         payload = "{'noRepoInfo': {'empty': 'https://github.com/foo/bar'}}".replace("'", "\"");
         GitHubTriggerProcessor gtp = new GitHubTriggerProcessor();
-        gtp.processGitHubPayload(payload, sbt.getClass());
+        gtp.trigger(payload);
```
Comment (jzoldak) [inline src/test/java/com/base2services/jenkins/GitHubTriggerProcessorTest.java]: Reviewer note:
This worked before because it was relying on the logic in the extractJsonFromPayload function that would have it return null if there was no 'repository' object in the payload. The new code changes will catch this case earlier, in the trigger function.

File: src/test/java/com/base2services/jenkins/ProcessCustomPayloadTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        repoSet.add(repoName);
+        doReturn(repoSet).when(sbt).getGitHubRepositories();
+    }
+
+    @Test
+    public void customMsgModeShouldPerformBuild() throws Exception {
+        payload = "{'job': 'testProject'}".replace("'", "\"");
+        GitHubTriggerProcessor gtp = new GitHubTriggerProcessor();
+        gtp.processCustomPayload(gtp.extractJsonFromPayload(payload), sbt.getClass());
+        // TODO verify that a job was scheduled.
```
Comment (jzoldak) [inline src/test/java/com/base2services/jenkins/ProcessCustomPayloadTest.java]: Reviewer note:
I'm not yet done with the tests for the custom message processing. I also plan to write a few tests for the trigger function, to verify its logic.

File: src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java [hunk 5]
```diff
@@ -71,8 +101,101 @@ public void processGitHubPayload(String payload, Class<? extends Trigger> trigge
...
+        // The custom payload must contain a root object named "job" of type string.
+        String jobToTrigger = json.getString("job");
+        if (jobToTrigger == null) {
+            LOGGER.warning("Custom sqs message payload does not contain information about which job to trigger.");
+            return;
+        }
+        // The custom payload can contain a root object named "parameters"
+        // that contains a list of parameters to pass to the job when scheduled.
+        // Each parameter object should contain information about its type, name, and value.
+        // TODO this will only work with string or boolean parameters,
```
Comment (benpatterson) [inline src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java]: Don't forget to add info to the jelly help file for this. Unless there is a generic "see wiki" reference, and this kind of detail goes onto the wiki...

File: src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java [hunk 6]
```diff
@@ -71,8 +101,101 @@ public void processGitHubPayload(String payload, Class<? extends Trigger> trigge
...
+        // and you need to pass in ALL parameters for the job,
+        // as the scheduled job will not fill in any defaults for you.
+        List<ParameterValue> parameters = getParamsFromJson(json);
+
+        Authentication old = SecurityContextHolder.getContext().getAuthentication();
```
Comment (benpatterson) [inline src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java]: why is it called `old`?

File: src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java [hunk 7]
```diff
@@ -71,8 +101,101 @@ public void processGitHubPayload(String payload, Class<? extends Trigger> trigge
...
+        SecurityContextHolder.getContext().setAuthentication(ACL.SYSTEM);
+        try {
```
Comment (benpatterson) [inline src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java]: I'm wondering if this needs an `except`? For example, if the job is not the correct job (line 126), or is not parameterized (line 128), etc.

File: src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java [hunk 8]
```diff
@@ -71,8 +101,101 @@ public void processGitHubPayload(String payload, Class<? extends Trigger> trigge
...
+        List<ParameterValue> params = new ArrayList<ParameterValue>();
+        if (json.has("parameters")) {
+            try {
+                JSONArray parameters = json.getJSONArray("parameters");
+                for (int i = 0; i < parameters.size(); i++) {
+                    JSONObject param = (JSONObject) parameters.get(i);
+                    if (isValidParameterJson(param)) {
+                        String name = param.getString("name");
+                        String type = param.getString("type");
+                        if (type.equals("boolean")) {
```
Comment (benpatterson) [inline src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java]: I might jsut put a comment in here (doesn't have to be this exact line) pointing out that the JSON string values are being cast as the ParameterValue objects appropriate for its type (e.g., boolean to BooleanParameterValue, etc). My memory isn't perfect with the `@link` annotation, but that might work in here as well. It's ultimately this object model that is forcing us to sort of hard-code to specific parameter types, if I'm understanding this correctly.

File: src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java (hunk 6 above)
Comment (jzoldak) [inline src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java]: That's a copy/paste from the original implementation. "old" is the original value, which is restored in the finally block.

File: src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java (hunk 7 above)
Comment (jzoldak) [inline src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java]: Those examples won't trigger an exception but instead will bypass the job for evaluation of whether or not it's the job you wanted. 

Hm, I could put in else blocks with logging messages. In the first case (jobName.equals...) that would print out a message for every jenkins job that doesn't match which will presumably be all the ones on jenkins BUT the one you wanted so that would be pretty noisy. Probably the best thing to do is to just add an else block for the (job instanceof ParameterizedJobMixIn.ParameterizedJob) clause. I'll do that.

File: src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java (hunk 5 above)
Comment (jzoldak) [inline src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java]: Added details to the help-manual.jelly file, this is the one that is shown when you click the ? button next to that radio button. I couldn't figure out a better place to put it - I'm definitely open to suggestions...

Comment (benpatterson) [inline src/main/java/com/base2services/jenkins/github/GitHubTriggerProcessor.java]: Reading through that now...No, I think what you have in the jelly file is great.

**Comment by jzoldak** [comment]
Closing this PR, submitted an upstream PR against the official jenkinsci repo.
